File: pincer/tools/edit.go
package tools

import (
	"fmt"
	"os"
	"strings"
)

type EditParams struct {
	Path      string `json:"path"`
	OldString string `json:"old_string"`
	NewString string `json:"new_string"`
}

func Edit(p *EditParams) (*Result, error) {
	data, err := os.ReadFile(p.Path)
	if err != nil {
		return &Result{Content: "Error reading file: " + err.Error(), IsError: true}, nil
	}

	content := string(data)
	count := strings.Count(content, p.OldString)

	if count == 0 {
		return &Result{
			Content: fmt.Sprintf("Error: old_string not found in %s", p.Path),
			IsError: true,
		}, nil
	}

	if count > 1 {
		return &Result{
			Content: fmt.Sprintf("Error: old_string found %d times in %s. It must be unique. Provide more context to make it unique.", count, p.Path),
			IsError: true,
		}, nil
	}

	newContent := strings.Replace(content, p.OldString, p.NewString, 1)
	if err := os.WriteFile(p.Path, []byte(newContent), 0644); err != nil {
		return &Result{Content: "Error writing file: " + err.Error(), IsError: true}, nil
	}

	return &Result{Content: fmt.Sprintf("Successfully edited %s", p.Path)}, nil
}

File: src/tools/edit.ts
import fs from "node:fs/promises";
import type { ToolResult } from "./bash.js";

export interface EditParams {
  path: string;
  old_string: string;
  new_string: string;
}

export async function edit(params: EditParams): Promise<ToolResult> {
  try {
    const content = await fs.readFile(params.path, "utf-8");

    const occurrences = content.split(params.old_string).length - 1;
    if (occurrences === 0) {
      return {
        content: `Error: old_string not found in ${params.path}`,
        isError: true,
      };
    }
    if (occurrences > 1) {
      return {
        content: `Error: old_string found ${occurrences} times in ${params.path}. It must be unique. Provide more context to make it unique.`,
        isError: true,
      };
    }

    const newContent = content.replace(params.old_string, params.new_string);
    await fs.writeFile(params.path, newContent, "utf-8");

    return { content: `Successfully edited ${params.path}` };
  } catch (err: any) {
    return {
      content: `Error editing file: ${err.message}`,
      isError: true,
    };
  }
}
